### ISSUE TITLE
Descriptor type being typed as "Unknown" after X assignments to it

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#3533]
Consider the following code:

```python
import enum


class SomeEnum(enum.Enum):
    FOO1 = "foo1"
    FOO2 = "foo2"
    FOO3 = "foo3"
    FOO4 = "foo4"
    FOO5 = "foo5"
    FOO6 = "foo6"
    FOO7 = "foo7"
    FOO8 = "foo8"
    FOO9 = "foo9"
    FOO10 = "foo10"
    FOO11 = "foo11"
    FOO12 = "foo12"
    FOO13 = "foo13"
    FOO14 = "foo14"
    FOO15 = "foo15"


class Descriptor:
    def __get__(self, obj, objtype=None) -> SomeEnum:
        ...

    def __set__(self, obj, value: SomeEnum):
        ...


class Foo:
    some_var = Descriptor()

    def some_method(self):
        self.some_var = SomeEnum.FOO1
        self.some_var = SomeEnum.FOO1
        self.some_var = SomeEnum.FOO2
        self.some_var = SomeEnum.FOO3
        self.some_var = SomeEnum.FOO4
        self.some_var = SomeEnum.FOO5
        self.some_var = SomeEnum.FOO6
        self.some_var = SomeEnum.FOO7
        self.some_var = SomeEnum.FOO8
        self.some_var = SomeEnum.FOO9
        self.some_var = SomeEnum.FOO10
        self.some_var = SomeEnum.FOO11
        self.some_var = SomeEnum.FOO12
        self.some_var = SomeEnum.FOO13
        self.some_var = SomeEnum.FOO14

    def other_method(self):
        reveal_type(self.some_var)


class Bar:
    some_var = Descriptor()

    def some_method(self):
        self.some_var = SomeEnum.FOO1
        self.some_var = SomeEnum.FOO1
        self.some_var = SomeEnum.FOO2
        self.some_var = SomeEnum.FOO3
        self.some_var = SomeEnum.FOO4
        self.some_var = SomeEnum.FOO5
        self.some_var = SomeEnum.FOO6
        self.some_var = SomeEnum.FOO7
        self.some_var = SomeEnum.FOO8
        self.some_var = SomeEnum.FOO9
        self.some_var = SomeEnum.FOO10
        self.some_var = SomeEnum.FOO11
        self.some_var = SomeEnum.FOO12
        self.some_var = SomeEnum.FOO13
        self.some_var = SomeEnum.FOO14
        self.some_var = SomeEnum.FOO15

    def other_method(self):
        reveal_type(self.some_var)
```

Running pyright will give:

```
❯ pyright xxx.py 
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/test.py
  /tmp/test.py:51:21 - information: Type of "self.some_var" is "SomeEnum"
  /tmp/test.py:76:21 - information: Type of "self.some_var" is "Unknown"
0 errors, 0 warnings, 2 informations 
Completed in 0.58sec
```

The difference between `Foo` and `Bar` is that `some_method` sets `some_var` 15 times for `Foo` and 16 times for `Bar`.

This is a minimal reproduction example and I created a single method setting all the possibilities at once, but a real world scenario would be an example on how I run into this issue: In a project of mine I have a descriptor column in a class with an enum that describes its state. That state had 15 possibilities set in 15 different methods and everything was fine. I just had to add one more possibility for the enum and suddenly its type started being identified as `Unknown`.

**Comment by bellini666** [comment]
obs. I just noticed that in the example above I set it to `FOO1` twice in a row by mistake, but even that triggered the issue.

I decided to test setting it to `FOO1` 15 times in a row for `Foo` and 16 times in a row for `Bar` and it also triggered the same issue. So, it doesn't matter if the var is being set to 16+ different possibilities, it will trigger if it is being set even for the same possibility 16+ times

**Comment by bellini666** [comment]
I see, that makes sense.

Regarding this use case, the reason why the descriptor is not annotated is because it is a django project, and it is not common design to annotate model fields in it. I just worked around this issue by doing exactly that, but I think it is an easy mistake to happen, specially when working with django or similar projects.

I'm fine if you decide to go with solution 1, but I would say that 2 seems more interesting just to have slightly more room for inferences (unless that would mean degrading the performance too much, of course)

**Comment by erictraut** [comment]
I've bumped the limit from 16 to 64. That matches the limit used for tuple, dict, set and list annotations.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.252, which I just published. It will also be included in the next release of pylance.
